components/api/src/main/java/com/hotels/styx/api/messages/StreamingHttpResponse.java

```diff
@@ -0,0 +1,448 @@
+import com.google.common.collect.ImmutableList;
+import com.google.common.net.MediaType;
+import com.hotels.styx.api.ContentOverflowException;
+import com.hotels.styx.api.HttpCookie;
+import com.hotels.styx.api.HttpHeaders;
+import com.hotels.styx.api.HttpResponse;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpVersion;
+import rx.Observable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.google.common.base.Objects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.hotels.styx.api.FlowControlDisableOperator.disableFlowControl;
+import static com.hotels.styx.api.HttpHeaderNames.CONTENT_LENGTH;
+import static com.hotels.styx.api.HttpHeaderNames.CONTENT_TYPE;
+import static com.hotels.styx.api.HttpHeaderNames.TRANSFER_ENCODING;
+import static com.hotels.styx.api.HttpHeaderValues.CHUNKED;
+import static io.netty.buffer.Unpooled.compositeBuffer;
+import static io.netty.handler.codec.http.HttpResponseStatus.OK;
+import static io.netty.handler.codec.http.HttpVersion.HTTP_1_1;
+import static io.netty.util.ReferenceCountUtil.release;
+import static java.lang.Integer.parseInt;
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * HTTP response with a fully aggregated/decoded body.
+ */
+public class StreamingHttpResponse implements StreamingHttpMessage {
+    private final HttpVersion version;
+    private final HttpResponseStatus status;
+    private final HttpHeaders headers;
+    private final Observable<ByteBuf> body;
+    private final List<HttpCookie> cookies;
+
+    StreamingHttpResponse(Builder builder) {
+        this.version = builder.version;
+        this.status = builder.status;
+        this.headers = builder.headers.build();
+        this.body = builder.body;
+        this.cookies = ImmutableList.copyOf(builder.cookies);
+    }
+
+    /**
+     * Creates an HTTP response builder with a status of 200 OK and empty body.
+     *
+     * @return a new builder
+     */
+    public static Builder response() {
+        return new Builder();
+    }
+
+    /**
+     * Creates an HTTP response builder with a given status and empty body.
+     *
+     * @param status response status
+     * @return a new builder
+     */
+    public static Builder response(HttpResponseStatus status) {
+        return new Builder(status);
+    }
+
+    /**
+     * Creates an HTTP response builder with a given status and body.
+     *
+     * @param status response status
+     * @param body response body
+     * @return a new builder
+     */
+    public static Builder response(HttpResponseStatus status, Observable<ByteBuf> body) {
```

Comment:
Is this a Netty `HttpResponseStatus`? We should aim to minimise Netty dependencies. Ideally use an `int` or introduce a Styx response status, but apart from ByteBuf let's keep this clean of Netty definitions.

Having said this I now realise this comment might also apply to FullHttpRequest/Response classes too.

Reply:
I will change the status as part of a separate branch as it will affect a very large number of files.